### DISCUSSION TITLE
Functional vs object oriented vs fluent API design

### DISCUSSION THREAD

**Discussion by axxel** [zxing-cpp/zxing-cpp#724] [General]
Working on the 3 most recent wrapper efforts (Rust, .NET, Kotlin/Native) made me realize that the `ReaderOptions` + `BarcodeReader` based API of some of the older wrappers is suboptimal / inconsistent.

Quick history recap: I replaced the overly complicated public API that was inherited from Java with the most trivial one I could think of: a single function call `ReadBarcode(image)` with some *optional* `ReaderOptions`. I saw no value in requiring to first instantiate some `Reader` object, only to then execute that read-function. That is still valid. But then came the wrapper subject where e.g. in Java or C# is it 100% idiomatic to model this not as a (static class) function but in an OO way.

Now I believe I was shortsighted when I decided to go with the `BarcodeReader` in Android+iOS but kept the separate `ReaderOptions` at the same time. That separation is superficial and adds no value but rather noise to the API. Bottom line: I'd like to change that and make sure each API/wrapper offers one (or both) of the following designs:
1. `ReadBarcodes` function + `ReaderOptions` struct
2. `BarcodeReader` class with all the options as properties, plus a `Barcodes read(ImageView)` function

Just in case that is not obvious: `BarcodeReader` is simply `ReaderOptions` with a `read()` function that passes `this` as the `options` argument to the underlying c++ call.

They may happily coexist in some environments (like C++) but others may choose to offer just one depending on the most natural approach of the platform (like Python, C, Android). And `BarcodeReader` may actually be derived from `ReaderOptions`, like in the [.NET wrapper](https://github.com/zxing-cpp/zxing-cpp/blob/4c5c0732c10f96b7ca5948684d26bcd6a0a7be91/wrappers/dotnet/ZXingCpp/ZXingCpp.cs#L316).

Since @markusfisch and @benjohnde are the most invested and contributed enormously to the Android and iOS code, what would you say if I went ahead with this backward incompatible change? I would then obviously do the full removal of the 3 deprecated options plus the `Result` -> `Barcode` renaming in the same release.

EDIT: for an extended discussion of the `fluent` part of the title, see [below](https://github.com/zxing-cpp/zxing-cpp/discussions/724#discussioncomment-8567239)

**Comment by markusfisch** [comment]
Hm, that would be quite a lot of changes at once 😉 Although I don't see any problems with the removal of the deprecated options and the `Result` -> `Barcode` renaming.

As you know, I'm not a fan of having an object just for the sake of it. I prefer so-called pure functions. So I really like the design of the "`ReadBarcodes` function + `ReaderOptions` struct".

But as far as I understand it, you would want to drop this design in Java/Kotlin in favor of the "`BarcodeReader` class with all the options as properties" design, because it's more idiomatic there, right?

Well, I'm not sure I agree with that since I am using static functions in Java all the time. And Kotlin even has top-level functions, which I use extensively, too. So I wouldn't say everything needs to be OO in Java or Kotlin. And the same is true for iOS, to some extent. Swift also has top-level functions, and you can, of course, have static functions in Objective-C, too. So for me, the "`ReadBarcodes` function + `ReaderOptions` struct" would still fit the bill in these languages as well. But I'm obviously biased.

My concern is that in the long run it doesn't make much sense to offer two ways of doing the same thing, mainly because there is no significant advantage of one approach over the other 😬

**Reply by axxel** [reply]
> Hm, that would be quite a lot of changes at once 😉 

Well, not in one commit, but in one release. If breaking the API is required than it is obviously better to break it once instead of 3 times, right?


> As you know, I'm not a fan of having an object just for the sake of it. I prefer so-called pure functions. So I really like the design of the "`ReadBarcodes` function + `ReaderOptions` struct".

I know. I remember looking into top-level functions in Kotlin last year but after learning how they are mapped into Java (I don't remember the details) I thought this is definitively *not* idiomatic. 

> But as far as I understand it, you would want to drop this design in Java/Kotlin in favor of the "`BarcodeReader` class with all the options as properties" design, because it's more idiomatic there, right?

I want to drop the inconsistent mix of those two designs that is currently there: `BarcodeReader` object + `BarcodeReader::Options` class + `BarcodeReader::read()` functions. The change I have in mind would actually trigger only one line of change in the current demo app code:
```patch
-					readerCpp.options.apply {
+					readerCpp.apply {
```

My point is that this 'options' entity adds no value.

> So for me, the "`ReadBarcodes` function + `ReaderOptions` struct" would still fit the bill in these languages as well. But I'm obviously biased.

I'm not against that. We can have both. But the current state is neither.

> My concern is that in the long run it doesn't make much sense to offer two ways of doing the same thing, mainly because there is no significant advantage of one approach over the other 😬

Yes, but maybe the advantage would be to make people with your preferences and those with others happy at the same time ;).

**Reply by markusfisch** [reply]
> Well, not in one commit, but in one release. If breaking the API is required than it is obviously better to break it once instead of 3 times, right?

Definitely, yes 👍 

> I know. I remember looking into top-level functions in Kotlin last year but after learning how they are mapped into Java (I don't remember the details) I thought this is definitively not idiomatic.

Well, for the JVM these top level functions end up being good old static functions 😉 And I use them all the time in Java. So I can't see anything wrong with that, but YMMV, of course.

> My point is that this 'options' entity adds no value.

With the current design, with `BarcodeReader` being an object, there's really not much value, agreed. Perhaps only that the reader options are better delineated. But that is of course a matter of taste.

Previously, `BarcodeReader` was a singleton, and having a separate `Options` class/struct made more sense there.

> Yes, but maybe the advantage would be to make people with your preferences and those with others happy at the same time ;).

Understood, and I like the ambition, of course 😉  I just think that it will become too much of a burden at some point, but of cource, I always prefer simplicity over developer convenience, but I know perfectly that isn't everyone's point of view 😉

**Comment by axxel** [comment]
Working on the writing part and contemplating the (a)symmetries between reading and writing, I came to the conclusion that taking the OO approach one step further and going full [fluent interface](https://en.wikipedia.org/wiki/Fluent_interface) style looks even more worthwhile and 'right' to me. With the plan to make `Barcode` the common class in both reading and writing, it became even more obvious that the counter part to reading is actually a 2 step process:

1. Text|Binary -> generate/create/build -> `Barcode`
2. `Barcode` -> write/render/... -> Image

This makes 'writing' symmetric to 'reading', as they both transform back and forth between an `Image` and a `Barcode`. But this requires a new operation to create a `Barcode` from a string or some binary data. And both the creation as well as the writing have a separate set of options. Zint has the same 2-step concept but combines all options in one struct.

So the current model symmetric to `ReaderOptions` would look something like
```c++
auto text = "Hello";
auto creatorOpts = CreatorOptions(BarcodeFormat::QRCode).setEcLevel("50%").setQuietZones(true);
auto barcode = CreateBarcodeFromTetx(text, creatorOpts);
auto writerOpts = WriterOptions().sizeHint(400).withQuietZones().withHRT().rotate(0);
auto image = WriteBarcodeToImage(barcode, writerOpts);
```
Note: `WriteBarcode` is not a good name as there is going to be at least a Bitmap and a SVG version.

This is already 'semi' fluent with respect to the creation of the `CreatorOptions` but going all the way would look like
```c++
using BarcodeFormat;
auto svg = Create(QRCode).withECLevel("50%").withQuietZones().fromText(text).asSVG();
```
or if we'd want to reuse the builder/creator:
```c++
auto createQRCode = Create(QRCode).withECLevel("50%").withQuietZones().rotatedBy(90).asGS1();
for (var text : inputs)
    std::cout << createQRCode.fromText(text).asSVG() << "\n";
```
This would also lend itself quite nicely in case we wanted to be explicit about the input, like `fromUtf8()` and `fromBinary()`.

Just to give an idea how this could then look for the reader side:
```c++
using BarcodeFormat;
auto iv = ImageView(...);
auto barcodes = Read(LinearCodes)
		.with(Binarizer::FixedThreshold)
		.with(TextMode::Plain)
		.withErrors()
		.withMinLineCount(3)
		.withSymbolLimit(2)
		.skipRotate()
		.fast()
		.from(iv);
		
// std::optional<Barcode> ReadPure::from()
auto pureBarcode = ReadPure(Pdf417).from(iv).value();
```
Notes:
 * separating `ReadPure()` seems logical as it is a rather separate from the rest with respect to a bunch of settings, like it only ever returns at most a single barcode, the default should be `setBinarizer(FixedThreshold)`, etc.
 * Instead of `Read` the function could be called `ReadBarcodes`
 * Instead of supplying the BarcodeFormat as a constructor parameter it could be passed like `Read().withFormats(Any)`
 * the names of the functions reflect their influence on the default setting, like `fast(bool v = true)` instead of `tryingHarder(bool v)`
 * the exact naming is tricky and up for debate
 * `readBarcodes.from(image)` looks a lot nicer to me than `reader.read(image)`
 * my idea is to have this API instead of a `BarcodeReader` / `BarcodeWriter`, the current `Barcodes ReadBarcodes(ImageView, ReaderOptions)` might very well stay.

Any suggestions with some motivation why this looks like a good/bad idea are welcome.

EDIT: updated the options naming of the writer part according to currently implemented experimental API.

**Reply by axxel** [reply]
Question for the encoding expert in the room (@gitlost): The zint API seems to imply that GS1 symbols (across all symbologies) can only ever contain "text", meaning ECI 899 can not be part of it. Can a GS1 symbol contain any ECI segment at all?

In the above outlined fluent interface design I can imagine using either of the following approaches:
 * something like `asGS1()` flag as in the example above, but that would allow calling `fromBinary()`
 * add a `fromGS1(string)` function, but that would allow setting a `withECI(ECI)` parameter
 * providing a separate `CreateGS1(BarcodeFormat)` factory that would not support setting any ECI/binary data, encoding this restriction in the type instead of throwing an error if configured inconsistently.

**Reply by isning** [reply]
Sorry, but I've just read this discussion, what do you think of the following fluent design?
In this way we can avoid strange 'class names' like 'Read', and it's also fluent for use.

```
using BarcodeFormat;
auto iv = ImageView(...);
auto barcodes = Barcode
		.from(iv)
		.with(Binarizer::FixedThreshold)
		.with(TextMode::Plain)
		.withErrors()
		.withMinLineCount(3)
		.withSymbolLimit(2)
		.skipRotate()
		.fast()
		.as(LinearCodes); // The name `as` is still up for debate
		
// std::optional<Barcode> ReadPure::from()
auto pureBarcode = Barcode.from(iv).as(Pdf417); // Different from Pdf417s?
// or this
auto pureBarcode = Barcode.from(iv).asPure(Pdf417);
```

**Reply by axxel** [reply]
Please note that `Read` in my above example is not a class name, but a factory function for an object of class `BarcodeReader`. But admittedly, from the users perspective, this is irrelevant and it looks like there is a class called `Read`. Your suggestion with `Barcode` will not work for different reasons: `Barcode` is a class, so `from` would need to be a static member function (in c++) and be called `Barcode::from(...)`. While that would be possible, I don't like it, because the return value is a list of barcodes. `Barcodes` also exists but is simply a `std::vector<Barcode>` which can not be extended. Then there is the issue that you kind of inverted the call order without explaining what types your functions return.

My idea was to have a `BarcodeReader` class, which needs to be created, then configured and finally used to scan the image data. But instead of using "traditional" naming like
```c++
auto reader = BarcodeReader();
reader.setFormats(LinearCodes);
reader.setReturnErrors(true);
reader.setTryHarder(false);
auto barcodes = reader.read(iv);
```
use a fluent API along the lines of
```c++
auto barcodes = Read(LinearCodes).withErrors().fast().from(iv);
```
Note that `fast()` returns a `BarcodeReader` object, while `from` returns a `Barcodes` object.

All that said, I understand why you found the name `from` on it's own on a `BarcodeReader` class non-intuitive. It only starts to make sens if you instanciate an object and name it with a verb like `auto readQRCodes = BarcodeReader(QRCode);` and then use it like `auto barcodes = readQRCodes.from(iv);`

**Reply by isning** [reply]
I agree with your opinion. I forgot that sometime we need to scan multiple barcodes. For me (maybe also for you), `Read(LinearCodes)` is a little strange. (It works and looks reasonable, but this design isn't commonly seen elsewhere, which may cause additional learning costs?)

And here I provide another two kind of design, hope it can inspire you or help you in any other way:
```
BarcodeReader.read(LinearCodes).withErrors().fast().from(iv)
// or
BarcodeReader().read(LinearCodes).withErrors().fast().from(iv)
```

If you have a more exact idea, please feel free to ask me to implement that for the kotlin wrapper, I will be happy to do that.

**Comment by gitlost** [comment]
Discounting CC-C composites (which use GS1-specific modified PDF417s),
there are only 3 GS1-approved symbologies that can have ECIs, all matrix,
namely GS1 Data Matrix, GS1 QR Code and GS1 DotCode.

Only for the last 2 are ECIs explicitly not supported (GS1 General
Specifications v24 Section 5.7.3 QR Code, Section 5.8.2 DotCode).

However, given that none of the allowed character sets can contain anything
other than subsets of ASCII, ECIs are useless anyway, and the lack of an
explicit statement about GS1 Data Matrix is probably just an omission.

GS1 have in the last few years introduced certain AIs, to do with transport
process (postal addresses), that specify using percent-encoding (RFC 3986)
for non-Latin characters (see Section 2.6.15).

Zint merely returns a warning if ECIs are used with GS1 data, for all 3 GS1
matrix symbologies, only returning an error if non-ASCII (or non-printable
ASCII) characters are used.

It also merely returns a warning if printable ASCII characters outside the
allowed GS1 character sets are used.

Re the fluent interface, I'm not really qualified to have an opinion. Are
errors not allowed? That would seem difficult to enforce...

**Comment by m7913d** [comment]
A functional, object oriented or fluent API are all fine to me. Object oriented API seems to me the most standard C++ approach. I prefer keeping backwards compatibility as much as possible.
